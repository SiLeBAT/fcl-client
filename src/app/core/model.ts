export interface ValidationError {
    code: string;
    message: string;
}
