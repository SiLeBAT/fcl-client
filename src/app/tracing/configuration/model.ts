import {
    Color,
    DataServiceInputState,
    LabelPart,
    NodeShapeType,
    TableColumn,
    TableRow,
} from "../data.model";
import { ComplexFilterCondition, PropToValuesMap } from "./configuration.model";

export interface RowFilter<T> {
    filter(arr: TableRow[]): TableRow[];
    getSettings(): T;
}

export interface TextFilterSettings {
    filterTerm: string;
    filterProps: string[];
}

export enum TableType {
    STATIONS,
    DELIVERIES,
}

export type TreeStatus = "collapsed" | "expanded";

export interface EditHighlightingServiceData {
    favouriteProperties: TableColumn[];
    otherProperties: TableColumn[];
    propToValuesMap: PropToValuesMap;
    ruleListItems: RuleListItem[];
}

export interface RuleListItem {
    id: RuleId;
    name: string;
    color: Color | null;
    shape: NodeShapeType | null;
    showInLegend: boolean;
    disabled: boolean;
    autoDisabled: boolean;
    isAnonymizationRule: boolean;
    effElementsCountTooltip: string;
    effElementsCount: number;
    conflictCount: number;
    ruleType: RuleType | null;
}

export interface EditRuleCore {
    id: string;
    name: string;
    userDisabled: boolean;
    complexFilterConditions: ComplexFilterCondition[];
    isValid: boolean;
}

export type EditRule =
    | ColorAndShapeEditRule
    | ColorEditRule
    | LabelEditRule
    | InvEditRule;

export enum RuleType {
    INVISIBILITY,
    LABEL,
    COLOR_AND_SHAPE,
    COLOR,
}

export interface ColorAndShapeEditRule extends EditRuleCore {
    type: RuleType.COLOR_AND_SHAPE;
    showInLegend: boolean;
    color: Color | null;
    shape: NodeShapeType | null;
}

export interface ColorEditRule extends EditRuleCore {
    type: RuleType.COLOR;
    showInLegend: boolean;
    color: Color;
}

export interface SimpleLabelEditRule extends EditRuleCore {
    type: RuleType.LABEL;
    labelProperty: string | null;
}

export interface ComposedLabelEditRule extends EditRuleCore {
    type: RuleType.LABEL;
    labelParts: LabelPart[];
    labelPrefix: string;
}

export type LabelEditRule = SimpleLabelEditRule | ComposedLabelEditRule;

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface InvEditRule extends EditRuleCore {
    type: RuleType.INVISIBILITY;
}

export type StationEditRule =
    | ColorAndShapeEditRule
    | LabelEditRule
    | InvEditRule;
export type DeliveryEditRule = LabelEditRule | InvEditRule | ColorEditRule;

export type DeliveryRuleType = RuleType.LABEL | RuleType.COLOR;
export type StationRuleType = RuleType.LABEL | RuleType.COLOR_AND_SHAPE;

export type RuleId = string;

export interface EditHighlightingState<T extends EditRule> {
    dataServiceInputState: DataServiceInputState;
    editRules: T[];
}
