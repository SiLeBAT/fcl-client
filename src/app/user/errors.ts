import { InvalidServerInputHttpErrorResponse } from '@app/core/errors';
import { ValidationError } from '@app/core/model';
import { CODE_TO_FIELD_MAP } from './consts/error-code-mappings.consts';
import { RegistrationCredentials } from './models/user.model';

export class InvalidRegistrationInput {
    private _errors: Partial<Record<keyof RegistrationCredentials, ValidationError[]>> = {};
    private _message: string | null = null;

    get errors(): Partial<Record<keyof RegistrationCredentials, ValidationError[]>> {
        return this._errors;
    }

    get message(): string | null {
        return this._message;
    }

    constructor(error: InvalidServerInputHttpErrorResponse) {
        this._message = error.message;
        this.mapErrors(error.errors);
    }

    private mapErrors(validationErrors: ValidationError[]): void {
        this._errors = {};
        validationErrors.forEach(error => {
            const field = CODE_TO_FIELD_MAP[error.code] as undefined | keyof RegistrationCredentials;
            if (field !== undefined) {
                const fieldErrors = this._errors[field];
                if (fieldErrors !== undefined) {
                    fieldErrors.push(error);
                } else {
                    this._errors[field] = [error];
                }
            }
        });
    }
}
