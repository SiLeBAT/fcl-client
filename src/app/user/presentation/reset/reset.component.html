<form [formGroup]="resetForm" (ngSubmit)="onReset()">
    <div class="fcl-mat-card-content">
        <div class="fcl-reset-text">
            <div class="mat-subheading-2">Please enter a new password.</div>
        </div>

        <div class="form-group" [ngClass]="{ error: !validateField('password1') }">
            <mat-form-field class="fcl-form-field" appearance="standard">
                <mat-label>New password</mat-label>
                <input matInput #message formControlName="password1" type="password" name="password1" required />
                <mat-icon matPrefix>lock</mat-icon>
                <mat-error *ngIf="!validateField('password1') && resetForm.controls.password1.errors.required">
                    Password is required
                </mat-error>
                <mat-error *ngIf="!validateField('password1') && resetForm.controls.password1.errors.serverValidationErrors">
                    <ng-container *ngFor="let error of resetForm.controls.password1.errors.serverValidationErrors; let i=index">
                        <br *ngIf="i>0"><span>{{ error.message }}</span>
                    </ng-container>
                </mat-error>
            </mat-form-field>
        </div>
        <div class="form-group">
            <mat-form-field class="fcl-form-field" appearance="standard">
                <mat-label>Confirm Password</mat-label>
                <input matInput formControlName="password2" type="password" name="password2" required />
                <mat-icon matPrefix>lock</mat-icon>
                <mat-error *ngIf="!validateField('password2') && resetForm.controls.password2.errors.validatePasswordConfirm">
                    Passwords must match
                </mat-error>
            </mat-form-field>
        </div>

        <div class="fcl-card-action fcl-card-action-container">
            <div class="form-group fcl-reset-button">
                <button mat-raised-button color="primary" type="submit" [disabled]="!resetForm.valid">
                    Recover password
                </button>
            </div>
            <div>
                <a mat-button [routerLink]="['/']">
                    Cancel
                </a>
            </div>
        </div>
    </div>
</form>
