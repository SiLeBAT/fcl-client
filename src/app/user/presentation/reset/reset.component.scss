@use "../../../../assets/sass/base" as fcl;

.fcl-card-action-container {
    display: flex;
}

.fcl-mat-card-content {
    width: 500px;
}

.fcl-reset-button {
    margin-right: fcl.$fcl-toolbar-padding * 4;
}

.fcl-reset-text,
.fcl-card-action {
    margin-top: 3rem;
}

.fcl-form-field {
    width: 100%;
}

em {
    @include fcl.em-errors;
}
