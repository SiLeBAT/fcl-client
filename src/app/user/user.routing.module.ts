import { NgModule } from "@angular/core";
import { Routes, RouterModule, mapToCanActivate } from "@angular/router";

import { LoginViewComponent } from "./presentation/login-view/login-view.component";
import { RegisterViewComponent } from "./presentation/register-view/register-view.component";
import { RecoveryViewComponent } from "./presentation/recovery-view/recovery-view.component";
import { ResetViewComponent } from "./presentation/reset-view/reset-view.component";
import { ActivateViewComponent } from "./presentation/activate-view/activate-view.component";
import { AdminActivateViewComponent } from "./presentation/admin-activate-view/admin-activate-view.component";
import { ProfileContainerComponent } from "./container/profile-container/profile-container.component";
import { environment } from "./../../environments/environment";
import { AuthGuard } from "./guards/auth.guard";
import { NewsActivateViewComponent } from "./presentation/news-activate-view/news-activate-view.component";

const userRoutes: Routes = [
    {
        path: "users",
        children: [
            {
                path: "profile",
                component: ProfileContainerComponent,
                canActivate: environment.serverless
                    ? undefined
                    : mapToCanActivate([AuthGuard]),
            },
            {
                path: "login",
                component: LoginViewComponent,
            },
            {
                path: "register",
                component: RegisterViewComponent,
            },
            {
                path: "recovery",
                component: RecoveryViewComponent,
            },
            {
                path: "reset/:id",
                component: ResetViewComponent,
            },
            {
                path: "activate/:id",
                component: ActivateViewComponent,
            },
            {
                path: "adminactivate/:id",
                component: AdminActivateViewComponent,
            },
            {
                path: "newsactivate/:id",
                component: NewsActivateViewComponent,
            },
        ],
    },
];

@NgModule({
    imports: [RouterModule.forChild(userRoutes)],
    exports: [RouterModule],
})
export class UserRoutingModule {}
