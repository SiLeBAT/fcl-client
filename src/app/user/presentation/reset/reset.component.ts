import { Component, OnInit, Output, EventEmitter, Input, ElementRef } from '@angular/core';
import { FormGroup, FormControl, Validators } from '@angular/forms';
import { ValidationError } from '@app/core/model';
import { NewPasswordRequestDTO } from '../../../user/models/user.model';

@Component({
    selector: 'fcl-reset',
    templateUrl: './reset.component.html',
    styleUrls: ['./reset.component.scss']
})
export class ResetComponent implements OnInit {

    @Input() serverValidationErrors: ValidationError[] = [];
    // eslint-disable-next-line @angular-eslint/no-output-native
    @Output() reset = new EventEmitter<NewPasswordRequestDTO>();

    resetForm: FormGroup;

    constructor(public elementRef: ElementRef) {}

    ngOnInit() {
        this.resetForm = new FormGroup({
            password1: new FormControl(null, [
                Validators.required
            ]),
            password2: new FormControl(null)
        }, this.passwordConfirmationValidator);
    }

    validateField(fieldName: string) {
        return this.resetForm.controls[fieldName].valid
      || this.resetForm.controls[fieldName].untouched;
    }

    onReset() {
        const password: NewPasswordRequestDTO = {
            password: this.resetForm.value.password1
        };
        this.reset.emit(password);
    }

    private passwordConfirmationValidator(fg: FormGroup) {
        const pw1 = fg.controls.password1;
        const pw2 = fg.controls.password2;

        if (pw1.value !== pw2.value) {
            pw2.setErrors({ validatePasswordConfirm: true });
        } else {
            pw2.setErrors(null);
        }
        return null;
    }
}
