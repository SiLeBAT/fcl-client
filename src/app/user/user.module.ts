import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { PasswordStrengthMeterModule } from 'angular-password-strength-meter';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';

import { SharedModule } from './../shared/shared.module';
import { CoreModule } from '../core/core.module';
import { MainPageModule } from '../main-page/main-page.module';
import { UserRoutingModule } from './user.routing.module';
import { ContentModule } from '@app/content/content.module';

import { STATE_SLICE_NAME, reducer } from './state/user.reducer';
import { UserEffects } from './state/user.effects';
import { LoginContainerComponent } from './container/login-container/login-container.component';
import { LoginComponent } from './presentation/login/login.component';
import { LoginViewComponent } from './presentation/login-view/login-view.component';
import { RegisterContainerComponent } from './container/register-container/register-container.component';
import { RegisterComponent } from './presentation/register/register.component';
import { RegisterViewComponent } from './presentation/register-view/register-view.component';
import { RecoveryContainerComponent } from './container/recovery-container/recovery-container.component';
import { RecoveryComponent } from './presentation/recovery/recovery.component';
import { RecoveryViewComponent } from './presentation/recovery-view/recovery-view.component';
import { ResetContainerComponent } from './container/reset-container/reset-container.component';
import { ResetComponent } from './presentation/reset/reset.component';
import { ResetViewComponent } from './presentation/reset-view/reset-view.component';
import { ActivateContainerComponent } from './container/activate-container/activate-container.component';
import { ActivateComponent } from './presentation/activate/activate.component';
import { ActivateViewComponent } from './presentation/activate-view/activate-view.component';
import { AdminActivateContainerComponent } from './container/admin-activate-container/admin-activate-container.component';
import { AdminActivateComponent } from './presentation/admin-activate/admin-activate.component';
import { AdminActivateViewComponent } from './presentation/admin-activate-view/admin-activate-view.component';
import { ProfileContainerComponent } from './container/profile-container/profile-container.component';
import { ProfileComponent } from './presentation/profile/profile.component';
import { GdprAgreementComponent } from './presentation/gdpr-agreement/gdpr-agreement.component';
import { NewsActivateContainerComponent } from './container/news-activate-container/news-activate-container.component';
import { NewsActivateComponent } from './presentation/news-activate/news-activate.component';
import { NewsActivateViewComponent } from './presentation/news-activate-view/news-activate-view.component';

@NgModule({
    imports: [
        CommonModule,
        ContentModule,
        ReactiveFormsModule,
        PasswordStrengthMeterModule,
        SharedModule,
        CoreModule,
        MainPageModule,
        UserRoutingModule,
        StoreModule.forFeature(STATE_SLICE_NAME, reducer),
        EffectsModule.forFeature([UserEffects]),
        FormsModule
    ],
    exports: [],
    declarations: [
        LoginContainerComponent,
        LoginComponent,
        LoginViewComponent,
        RegisterContainerComponent,
        RegisterComponent,
        RegisterViewComponent,
        RecoveryViewComponent,
        RecoveryComponent,
        RecoveryContainerComponent,
        ResetContainerComponent,
        ResetComponent,
        ResetViewComponent,
        ActivateContainerComponent,
        ActivateComponent,
        ActivateViewComponent,
        AdminActivateContainerComponent,
        AdminActivateComponent,
        AdminActivateViewComponent,
        ProfileContainerComponent,
        ProfileComponent,
        GdprAgreementComponent,
        NewsActivateContainerComponent,
        NewsActivateComponent,
        NewsActivateViewComponent
    ],
    schemas: [CUSTOM_ELEMENTS_SCHEMA]

})
export class UserModule { }
