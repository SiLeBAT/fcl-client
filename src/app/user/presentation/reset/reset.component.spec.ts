import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { ResetComponent } from "./reset.component";
import { MaterialModule } from "../../../shared/material.module";
import { SharedModule } from "../../../shared/shared.module";
import { NewPasswordRequestDTO } from "../../models/user.model";
import { RouterTestingModule } from "@angular/router/testing";
import { NoopAnimationsModule } from "@angular/platform-browser/animations";

describe("ResetComponent", () => {
    let component: ResetComponent;
    let fixture: ComponentFixture<ResetComponent>;

    beforeEach(async(() => {
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        TestBed.configureTestingModule({
            declarations: [ResetComponent],
            imports: [
                MaterialModule,
                SharedModule,
                RouterTestingModule,
                NoopAnimationsModule,
            ],
        })
            .compileComponents()
            .then(() => {
                fixture = TestBed.createComponent(ResetComponent);
                component = fixture.componentInstance;
            });
    }));

    it("should create the component", () => {
        expect(component).toBeTruthy();
    });

    it("should render the component correctly", () => {
        expect(fixture).toMatchSnapshot();
    });

    it("should have an invalid reset form when empty", () => {
        expect(component.resetForm.valid).toBeFalsy();
    });

    it("should have an invalid password field when being empty", () => {
        const password1Field = component.resetForm.controls["password1"];

        expect(password1Field.valid).toBeFalsy();
    });

    it("should have an required validator on the password field", () => {
        let errors = {};
        const password1Field = component.resetForm.controls["password1"];
        errors = password1Field.errors || {};

        expect(errors["required"]).toBeTruthy();
    });

    it("should have an invalid status with two different passwords", () => {
        const password1Field = component.resetForm.controls["password1"];
        const password2Field = component.resetForm.controls["password2"];
        password1Field.setValue("test");
        password2Field.setValue("testt");
        const errors = password2Field.errors || {};

        expect(errors["validatePasswordConfirm"]).toBeTruthy();
    });

    it("should emit the email when submitting the recovery form", () => {
        const password = "testtest";

        expect(component.resetForm.valid).toBeFalsy();

        component.resetForm.controls["password1"].setValue(password);
        component.resetForm.controls["password2"].setValue(password);

        expect(component.resetForm.valid).toBeTruthy();

        let passwordRequest: NewPasswordRequestDTO;
        component.reset.subscribe((value) => (passwordRequest = value));

        component.onReset();

        expect(passwordRequest!.password).toBe(password);
    });
});
