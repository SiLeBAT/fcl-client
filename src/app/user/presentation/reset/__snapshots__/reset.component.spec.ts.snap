// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetComponent should render the component correctly 1`] = `
<fcl-reset
  changeRef={[Function ViewRef_]}
  pwStrength={[Function Number]}
  reset={[Function EventEmitter_]}
  resetForm={[Function FormGroup]}
>
  <form
    novalidate=""
  >
    <div
      class="fcl-mat-card-content"
    >
      <div
        class="fcl-reset-text"
      >
        <div
          class="mat-subheading-2"
        >
          Please enter a new password.
        </div>
      </div>
      <div
        class="form-group"
      >
        <mat-form-field
          appearance="standard"
          class="fcl-form-field mat-form-field mat-primary"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  class="mat-input-element mat-form-field-autofill-control"
                  formcontrolname="password1"
                  matinput=""
                  name="password1"
                  required=""
                  type="password"
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-subscript-wrapper"
            >
              
              
            </div>
          </div>
        </mat-form-field>
      </div>
      <div
        class="form-group"
      >
        <password-strength-meter>
          <div>
            <div
              class="strength-meter"
            >
              <div
                class="strength-meter-fill"
              />
            </div>
            
          </div>
        </password-strength-meter>
        
      </div>
      <div
        class="form-group"
      >
        <mat-form-field
          appearance="standard"
          class="fcl-form-field mat-form-field mat-primary"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  class="mat-input-element mat-form-field-autofill-control"
                  formcontrolname="password2"
                  matinput=""
                  name="password2"
                  required=""
                  type="password"
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-subscript-wrapper"
            >
              
              
            </div>
          </div>
        </mat-form-field>
      </div>
      <div
        class="fcl-card-action fcl-card-action-container"
      >
        <div
          class="form-group fcl-reset-button"
        >
          <button
            class="mat-focus-indicator mat-raised-button mat-button-base"
            color="primary"
            mat-raised-button=""
            type="submit"
          >
            <span
              class="mat-button-wrapper"
            >
               Recover password 
            </span>
            <span
              class="mat-button-ripple mat-ripple"
              matripple=""
            />
            <span
              class="mat-button-focus-overlay"
            />
          </button>
        </div>
        <div>
          <a
            class="mat-focus-indicator mat-button mat-button-base"
            mat-button=""
          >
            <span
              class="mat-button-wrapper"
            >
               Cancel 
            </span>
            <span
              class="mat-button-ripple mat-ripple"
              matripple=""
            />
            <span
              class="mat-button-focus-overlay"
            />
          </a>
        </div>
      </div>
    </div>
  </form>
</fcl-reset>
`;
